README.md
# automationLibs_logger Library

automationLibs is a collection of Python packages that are created by ISG's Automation Squad. automationLibs aims to
create boilerplate functions that will allow you to start coding immediately. Here is a list of other libraries:

* [automationLibs_logger Library][1] - Logger Library
* automationLibs_Gdrive Library - GDrive and GSheet helper Library
* automationLibs_PyAutoGUI Library - PyAutoGUI related Library
* automationLibs_Web Library - Selenium and other web related Library
* automationLibs_remote Library - sftp Library
* automationLibs_service Library - other helper Libraries

[1]:https://gitlab.com/yasmin.barrientos/automationLibs_logger

## Installation and updating

1. Create a .ENV in your root folder
2. Add full path where you would like the logs to be created(could be anywhere:
```dotenv
#*** automationLibs_logger ***#
LOG_BASE_DIR="C:\\Users\\my_machine\\PycharmProjects\\my_project\\logs"
```

3. Use the package manager [pip](https://pip.pypa.io/en/stable/) to install automationLibs like below. Rerun this command
to check for and install updates .

```bash
pip install -e git+https://yasmin.barrientos/automationLibs_logger#egg=automationLibs_logger
```

To install a specific branch

```bash
pip install -e git+https://gitlab.com/yasmin.barrientos/automationLibs_logger.git@dev-ver01#egg=automationlibs-logger
```

## Usage

Features:
Install this package to use a pre-formatted logger

##### Logger

* Logger(LOG_NAME_ID).logger
    * instantiates and creates the logger

#### Demo of logger:

```python
from loggerLibrary.logger import Logger
import getpass

LOG_NAME_ID = f'loggerLibrarytest_{getpass.getuser()}'
logger = Logger(LOG_NAME_ID).logger

logger.info('test info')
logger.error('test info')
```

Result in PyCharm Terminal:

```
[2022-03-28 17:56:30] : [INFO] : [C:\DEV\PYTHON\PycharmProjects\loggerLibrary\tests\test_logger.py:9] : test info;
[2022-03-28 17:56:30] : [ERROR] : [C:\DEV\PYTHON\PycharmProjects\loggerLibrary\tests\test_logger.py:10] : test info;
```

Result in Log File:

```log
[2022-03-28 17:56:30] : [INFO] : [C:\DEV\PYTHON\PycharmProjects\loggerLibrary\tests\test_logger.py:9] : test info;
[2022-03-28 17:56:30] : [ERROR] : [C:\DEV\PYTHON\PycharmProjects\loggerLibrary\tests\test_logger.py:10] : test info;
```

#### Demo of Old Logger:

```python
from loggerLibrary.old_logger import OldLogger
import getpass

LOG_NAME_ID = f'loggerLibrarytest_{getpass.getuser()}'
logger2 = OldLogger(log_dir=2, log_name_id=LOG_NAME_ID).logger()

logger2.info('test info')
logger2.error('test info')

```

Result in PyCharm Terminal:

```log
[2022-03-28 17:56:30]: [INFO]: [test_logger.py:14] : test info;
[2022-03-28 17:56:30]: [ERROR]: [test_logger.py:15] : test info;
```

Result in Log File:

```log
[2022-03-28 17:56:30]: [INFO]: [C:\DEV\PYTHON\PycharmProjects\loggerLibrary\tests\test_logger.py:14] : test info;
[2022-03-28 17:56:30]: [ERROR]: [C:\DEV\PYTHON\PycharmProjects\loggerLibrary\tests\test_logger.py:15] : test info;
```

## Log Format
For more log formats, check here: https://docs.python.org/3/library/logging.html#logrecord-attributes

## License

[MIT](https://choosealicense.com/licenses/mit/)


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file.
 
The format is based on [Keep a Changelog](http://keepachangelog.com/)
and this project adheres to [Semantic Versioning](http://semver.org/).
 
## [Unreleased] - yyyy-mm-dd
 
Here we write upgrading notes for brands. It's a team effort to make them as
straightforward as possible.
 
### Initial Commit
- [loggerLibrary-2022-03-28](http://tickets.projectname.com/browse/PROJECTNAME-XXXX)
  - Initial commit of Logger.
  - Added copy of old logger for reference
 
### Changed
 
### Fixed
 
## [1.2.4] - 2017-03-15
  
Here we would have the update steps for 1.2.4 for people to follow.
 
### Added
 
### Changed
  
- [PROJECTNAME-ZZZZ](http://tickets.projectname.com/browse/PROJECTNAME-ZZZZ)
  PATCH Drupal.org is now used for composer.
 
### Fixed
 
- [PROJECTNAME-TTTT](http://tickets.projectname.com/browse/PROJECTNAME-TTTT)
  PATCH Add logic to runsheet teaser delete to delete corresponding
  schedule cards.
 
## [1.2.3] - 2017-03-14
 
### Added
   
### Changed
 
### Fixed
 
- [PROJECTNAME-UUUU](http://tickets.projectname.com/browse/PROJECTNAME-UUUU)
  MINOR Fix module foo tests
- [PROJECTNAME-RRRR](http://tickets.projectname.com/browse/PROJECTNAME-RRRR)
  MAJOR Module foo's timeline uses the browser timezone for date resolution 

loggerLibrary/logger.py
import pathlib
import logging
import os
from datetime import datetime


class ContextFilter(logging.Filter):
    """
    This is a filter which injects contextual information into the log.
    """

    def filter(self, record):
        record.today = datetime.today().strftime('%Y-%m-%d')
        record.time = datetime.today().strftime('%H:%M:%S')
        return True


class Logger:
    """
        Logger class that sets up python logger
        :params str app_name: name of the app, would be part of the file name of the log file.
        :params str set_logger_name: set the name of the app to logger name, so that you can choose which file to log if
         you have multiple log files.

        Sample use 1:
        logger = Logger(app_name).logger

        Sample use 2:
        import logging

        Logger(app_name_a, True)
        Logger(app_name_b, True)

        logger_a = logging.getLogger("app_name_a")
        logger_b = logging.getLogger("app_name_b")

        logger_a.info("Hello")
        logger_b.info("Hi")
    """

    def __init__(self, app_name, set_logger_name: bool = False) -> None:
        self.base_dir = os.environ.get('LOG_BASE_DIR') or os.path.dirname(os.path.dirname(os.path.abspath(__file__)))
        self.app_name = app_name
        self.set_logger_name = set_logger_name
        self.LOG_DIR = os.path.join(self.base_dir, 'logs', datetime.today().strftime('%Y'),
                                    datetime.today().strftime('%B'))
        self.with_milliseconds = True

        self.logger = self.make_logger()

    def make_logger(self):
        # checks if directory exists, if not create it
        pathlib.Path(self.LOG_DIR).mkdir(parents=True, exist_ok=True)

        self.LOG_DIR = os.path.join(self.LOG_DIR,
                                    self.app_name + '_log_' + datetime.today().strftime('%Y-%m-%d') + '.log')

        if self.set_logger_name:
            logger = logging.getLogger(self.app_name)
        else:
            logger = logging.getLogger(__name__)
        logger.setLevel(logging.DEBUG)

        if self.with_milliseconds:
            formatter = logging.Formatter('[%(asctime)s] : [%(levelname)s] : [%(pathname)s:%(lineno)d] : %(message)s;')
        else:
            formatter = logging.Formatter('[%(asctime)s] : [%(levelname)s] : [%(pathname)s:%(lineno)d] : %(message)s;',
                                          '%Y-%m-%d %H:%M:%S')

        file_handler = logging.FileHandler(self.LOG_DIR)
        file_handler.setFormatter(formatter)
        logger.addHandler(file_handler)
        logger.addFilter(ContextFilter())

        # create console handler and set level to debug
        ch = logging.StreamHandler()
        ch.setLevel(logging.DEBUG)

        # add formatter to ch
        ch.setFormatter(formatter)

        # add ch to logger
        logger.addHandler(ch)

        return logger


loggerLibrary/read_logs.py
from datetime import date, datetime, timedelta
from . import log_config
from os.path import exists, join


class ReadLogs:

    def help(self):
        return ("""
        Function that returns data from log files
        :param date: accepts date value (yyyy-mm-dd) and looks for log file with that date
        :param *args: accepts lists of important words separated by comma, function will find line items with to those words in it
        """)

    def read_log(self, date, keep_phrases=None):
        """Function that returns data from log files
        :param date: accepts date value (yyyy-mm-dd) and looks for log file with that date
        :param *args: accepts lists of important words separated by comma, function will find line items with to those words in it
        """
        try:
            month = datetime.strptime(date[5:7], "%m")
            path_exists = exists(join(log_config.BASE_DIR, 'logs/', date[:-6], month.strftime('%B')))
            print(join(log_config.BASE_DIR, 'logs/', date[:-6], date[5:7]))
            print(path_exists)
            important = []
            if path_exists:
                file = join(log_config.CURRENT_LOG_DIR, log_config.APP_NAME + '_log_' + date + '.log')
                print('keep phrases: ', keep_phrases)
                with open(file) as f:
                    f = f.readlines()

                for line in f:
                    if keep_phrases:
                        for phrase in keep_phrases:
                            if phrase in line:
                                important.append(line)
                                break
                    else:
                        important.append(line)

            print(' '.join([str(elem) for elem in important]))
            return important

        except Exception as e:
            log_config.logger.error(e)

    def filter_range(self, start, end, time=None, keep_phrases=None):
        """
            yung search pwede syang maglagay ng date range, 
            time range, para mafilter nya as a support yung kelangan nya lang kunin, 
            baka mamaya kasi sobrang dami and irelevant na sa kelangan nyan date or time
        """

        # for time filter, we need to parse line, split on ':', get last index then split again for " " then look for first 2 digits
        important = []

        try:
            for single_date in self.daterange(start, end):
                print(single_date.strftime("%Y-%m-%d"))
                # path_exists = exists(join(log_config.BASE_DIR, 'logs/', date[:-3], date[:-3]))
                CURRENT_LOG_DIR = join(log_config.BASE_DIR, 'logs', single_date.strftime('%Y'),
                                       single_date.strftime('%B'))
                file = join(CURRENT_LOG_DIR, log_config.APP_NAME + '_log_' + single_date.strftime("%Y-%m-%d") + '.log')

                print('curr file is: ', file)
                with open(file) as f:
                    f = f.readlines()

                for line in f:
                    if time and self.get_time(time, line):
                        important.append(line)

                    if keep_phrases:
                        for phrase in keep_phrases:
                            if phrase in line:
                                important.append(line)
                                break
                    # else:
                    # important.append(line)

            print(' '.join([str(elem) for elem in important]))
            return important

        except Exception as e:
            print(e)

        return important

    @staticmethod
    def daterange(start_date, end_date):
        format = '%Y-%m-%d'
        start_date, end_date = datetime.strptime(start_date, format), datetime.strptime(end_date, format)
        for n in range(int((end_date - start_date).days) + 1):
            yield start_date + timedelta(n)

    @staticmethod
    def get_time(time, line):
        datetime = line.split(': ')
        time_in_line = datetime[len(datetime) - 2]

        if time in time_in_line.split(':')[0]:
            return True

        return False


loggerLibrary/log_config.py
import pathlib
import logging
import os
from datetime import datetime

class ContextFilter(logging.Filter):
    """
    This is a filter which injects contextual information into the log.
    """
    def filter(self, record):
        record.today = datetime.today().strftime('%Y-%m-%d')
        record.time = datetime.today().strftime('%H:%M:%S')
        return True
        

BASE_DIR = os.environ.get('LOG_BASE_DIR') or os.path.dirname(os.path.dirname(os.path.abspath(__file__)))
APP_NAME = "logger_library"
CURRENT_LOG_DIR = os.path.join(BASE_DIR, 'logs', datetime.today().strftime('%Y'), datetime.today().strftime('%B'))

#checks if directory exists, if not create it
pathlib.Path(CURRENT_LOG_DIR).mkdir(parents=True, exist_ok=True) 

LOG_DIR = os.path.join(CURRENT_LOG_DIR,  APP_NAME + '_log_' + datetime.today().strftime('%Y-%m-%d') + '.log')


logger = logging.getLogger(__name__)
logger.setLevel(logging.DEBUG)
formatter = logging.Formatter('[%(asctime)s] : [%(levelname)s] : %(pathname)s : %(message)s;')
file_handler = logging.FileHandler(LOG_DIR)
file_handler.setFormatter(formatter)
logger.addHandler(file_handler)
logger.addFilter(ContextFilter())

# create console handler and set level to debug
ch = logging.StreamHandler()
ch.setLevel(logging.DEBUG)

# add formatter to ch
ch.setFormatter(formatter)

# add ch to logger
logger.addHandler(ch)






loggerLibrary/old_logger.py
import pathlib
import logging
import os
from datetime import datetime


class OldLogger:
    def __init__(self, log_dir=1, log_name_id=None):
        """
        For reference: https://docs.python.org/3/library/logging.html
        :param log_dir: Complete folder location, where the log file will be created. If 1 or None: Logs will be created
        in Logs folder. If 2: logs will be created in Logs/<year>/<Month> Folder, e.g. logs/2022/March
        :type log_dir: int
        :param log_name_id: identifying name of the log. Could be based on the app name and/or machine to make the log
            file easier to identify. e.g. myLog_myMachine --> myLog_myMachine_log_2022-03-28.log.
            Otherwise, log file will be named: log_2022-03-28.log
        :type log_name_id: string
        """
        self.base_dir = os.environ.get('LOG_BASE_DIR') or os.path.dirname(os.path.dirname(os.path.abspath(__file__)))

        if log_dir == 1 or log_dir is None:
            self.LOG_DIR = os.path.join(self.base_dir, 'logs')
        elif log_dir == 2:
            self.LOG_DIR = os.path.join(self.base_dir,
                                        'logs',
                                        datetime.today().strftime('%Y'),
                                        datetime.today().strftime('%B'))
        else:
            self.LOG_DIR = log_dir

        if log_name_id is None:
            self.LOG_DIR = os.path.join(self.LOG_DIR,
                                        'log_' + datetime.today().strftime('%Y-%m-%d') + '.log')
        else:
            self.LOG_DIR = os.path.join(self.LOG_DIR,
                                        log_name_id + '_log_' + datetime.today().strftime('%Y-%m-%d') + '.log')

    def logger(self):
        # log format in log file
        # for more formats, check here: https://docs.python.org/3/library/logging.html#logrecord-attributes
        logging.basicConfig(filename=self.LOG_DIR,
                            filemode='a',
                            level=logging.DEBUG,
                            format='[%(asctime)s]: [%(levelname)s]: [%(pathname)s:%(lineno)d] : %(message)s;',
                            datefmt='%Y-%m-%d %H:%M:%S')

        # create logger
        logger = logging.getLogger(self.LOG_DIR)
        logger.setLevel(logging.DEBUG)

        # create console handler and set level to debug
        ch = logging.StreamHandler()
        ch.setLevel(logging.DEBUG)

        # log format in PyCharm terminal
        # for more formats, check here: https://docs.python.org/3/library/logging.html#logrecord-attributes
        formatter = logging.Formatter('[%(asctime)s]: [%(levelname)s]: [%(pathname)s:%(lineno)d] : %(message)s;',
                                      '%Y-%m-%d %H:%M:%S')

        # add formatter to ch
        ch.setFormatter(formatter)

        # add ch to logger
        logger.addHandler(ch)

        return logger
